numpy/_core/src/umath/string_ufuncs.cpp

```diff
@@ -947,6 +998,54 @@ string_zfill_resolve_descriptors(
 }
 
 
+static int
+string_partition_promoter(PyObject *NPY_UNUSED(ufunc),
+        PyArray_DTypeMeta *const op_dtypes[], PyArray_DTypeMeta *const signature[],
+        PyArray_DTypeMeta *new_op_dtypes[])
+{
+    Py_INCREF(op_dtypes[0]);
+    new_op_dtypes[0] = op_dtypes[0];
+    Py_INCREF(op_dtypes[1]);
+    new_op_dtypes[1] = op_dtypes[1];
+
+    new_op_dtypes[2] = NPY_DT_NewRef(&PyArray_Int64DType);
+
+    Py_INCREF(op_dtypes[3]);
```

Comment:
It's not clear to me why it hasn't been a problem until this iteration of the pull request (perhaps a change on main?) but the problem on 32 bit architectures is coming from this line. It's actually not anything special about 32 bit, only in that the default int is 32 bit there so we trigger this promoter. You can trigger the same bug on on 64 bit with the following script:

```
import numpy as np

buf = np.array("this is a string to partition", dtype="S")
sep = np.array("th", dtype="S")
idx = np.array(0, dtype='int32')
out1 = np.array("", dtype="S32")
out2 = np.array("", dtype="S32")
out3 = np.array("", dtype="S32")

print(np._core.umath._partition_index(buf, sep, idx, out=(out1, out2, out3)))
```

Not clear to me if this is an issue with the promoters in general or an issue with this promoter implementation in particular.

Reply:
Is that fixed by #26116?  See https://github.com/numpy/numpy/pull/26116#issuecomment-2016519340 in particular.

Reply:
What's happening is the dtype is explicitly NULLed here, there's a note about how outputs are ignored for promoters if they're not in the signature:

https://github.com/numpy/numpy/blame/c3694335faffe35e3d961feffefdc0a4b9a665da/numpy/_core/src/umath/dispatching.c#L964

In my promoter implementations in stringdtype I avoid this issue since I can just say e.g. `new_op_dtypes[2] = NPY_DT_NewRef(&PyArray_StringDType);`. That doesn't work for the unicode and bytes dtype, because the promoter doesn't know which dtype it's being called for. It looks like there are a number of other promoters for bytes and unicode that have the same issue as is being triggered here.

Reply:
It is confusing. I guess one should just set the output ones to `NULL`?

Independently, probably good to add the test case you have to the tests!

Reply:
Yeah maybe just replacing INCREF with XINCREF is sufficient, but I was hoping @seberg could clarify. And we should probably fix the other promoters in this file similarly. I’ll leave that to Lysandros.

Reply:
And for a test rather than just pasting what I have above, we probably want to systematically test the promoters we add by adding tests that trigger them.

Also see my suggestion above about how we can probably get rid of the promoters for these ufuncs by casting to e.g. int64 in the python wrapper.

Reply:
SGTM